sns/parity_math.c

```diff
@@ -1426,14 +1442,51 @@ M0_INTERNAL void m0_parity_math_buffer_xor(struct m0_buf *dest,
 			(m0_parity_elem_t)((uint8_t*)src[0].b_addr)[ei];
 }
 
+#if ISAL_ENCODE_ENABLED
+static int isal_ir_init(struct m0_sns_ir *ir)
+{
+	int ret;
+
+	M0_PRE(ir != NULL);
+
+	/* Allocate memory to hold indices of alive blocks */
+	ALLOC_ARR_INFO(ir->si_alive_idx, ir->si_alive_nr,
+		       "alive index buffer", ret);
+	if (ir->si_alive_idx == NULL)
+		return ret;
+
+	/* Allocate memory to hold indices of failed blocks */
+	ALLOC_ARR_INFO(ir->si_failed_idx, ir->si_parity_nr,
+		       "failed index buffer", ret);
+	if (ir->si_failed_idx == NULL)
+		return ret;
```

Comment:
Here, we need to free "ir->si_alive_idx" before returning error.

Reply:
After returning from here we are calling m0_sns_ir_fini() in case of error. So there we are freeing all buffers including ir->si_alive_idx.

Reply:
The basic rule of function behavior is that: if a function returns error, it should undo all things before return.

In this case, if this function returns error, it should free all memory previously allocated in this function.
And if xxx_init() fails, xxx_fini() should not be called.
xxx_fini() is only called when xxx_init() succeeded.

Reply:
I agree. Then we need to update m0_parity_math_init() as well because it is calling m0_parity_math_fini() to handle error.

Reply:
OK.